src/features/pass-survey/model/hooks/useActivityState.ts

```diff
@@ -0,0 +1,133 @@
+import { useMemo } from 'react';
+
+import { ActivityDto } from '@app/shared/api';
+import { ActivityModel, useActivityDetailsQuery } from '@entities/activity';
+
+import { useActivityStorage } from '../../lib';
+import { buildPipeline } from '../pipelineBuilder';
+
+type UseActivityPipelineArgs = {
+  appletId: string;
+  activityId: string;
+  eventId: string;
+};
+
+const mockActivity: ActivityDto = {
+  id: 'aid1',
+  name: 'Activity number 1',
+  description:
+    'Activity description number 1 Activity description 1 number 1 Activity description number 1',
+  image:
+    'https://raw.githubusercontent.com/mtg137/Stability_tracker_applet/master/protocols/stability/mindlogger-logo.png',
+  splashScreen: '',
+  showAllAtOnce: false,
+  isSkippable: false,
+  isReviewable: false,
+  responseIsEditable: false,
+  ordering: 0,
+  items: [
+    {
+      id: 100,
+      inputType: 'AbTest',
+      config: {
+        device: 'Phone',
+      },
+      timer: 0,
+      hasTokenValue: true,
+      isSkippable: true,
+      hasAlert: true,
+      hasScore: true,
+      isAbleToMoveToPrevious: true,
+      hasTextResponse: true,
+      order: 0,
+    },
+  ],
+};
+
+function useActivityState({
+  appletId,
+  activityId,
+  eventId,
+}: UseActivityPipelineArgs) {
+  const { activityStorage, updateActivityStorage } = useActivityStorage({
+    appletId,
+    activityId,
+    eventId,
+  });
+
+  let { data: activity } = useActivityDetailsQuery(activityId, {
+    enabled: !activityStorage,
+    select: r => r.data.result,
+  });
+
+  // @todo remove once integration is done
+  if (!activity) {
+    activity = mockActivity;
+  }
+
+  const pipeline = useMemo(
+    () =>
+      activity ? buildPipeline(ActivityModel.mapToActivity(activity)) : [],
+    [activity],
+  );
+
+  const shouldCreateStorage = !activityStorage && pipeline.length;
+
+  if (shouldCreateStorage) {
+    createActivityStorage();
+  }
+
+  function createActivityStorage() {
+    updateActivityStorage({
```

Comment:
Maybe upsertActivityStorage = update / insert  - often used on backend for similar cases